editor/src/clj/editor/model_scene.clj

```diff
@@ -310,86 +256,173 @@
+                                   (geom/aabb-transform aabb transform)))
+                            geom/aabb-union
+                            geom/null-aabb
+                            renderable-models)]
+        {:aabb mesh-set-aabb
+         :renderable-models renderable-models}))))
+
+(g/defnk produce-renderable-mesh-set [content]
+  (let [mesh-material-index->material-name
+        (or (some-> content :material-ids not-empty vec)
+            default-material-ids)]
+    (make-renderable-mesh-set (:mesh-set content) mesh-material-index->material-name)))
+
+(defn- make-mesh-scene [renderable-mesh model-scene-resource-node-id]
+  (let [{:keys [aabb material-name renderable-data]} renderable-mesh]
+    {:node-id model-scene-resource-node-id
+     :aabb aabb
+     :renderable {:render-fn render-mesh
+                  :tags #{:model}
+                  :batch-key nil ; Batching is disabled in the editor for simplicity.
+                  :select-batch-key model-scene-resource-node-id
+                  :passes [pass/opaque pass/opaque-selection]
+                  :user-data {:mesh-renderable-data renderable-data
+                              :vertex-space :vertex-space-local
+                              :material-name material-name
+                              :shader preview-shader
+                              :textures {"texture_sampler" @texture/white-pixel}}}}))
+
+(defn- make-model-scene [renderable-model model-scene-resource-node-id]
+  (let [{:keys [transform aabb renderable-meshes]} renderable-model
+        mesh-scenes (mapv #(make-mesh-scene % model-scene-resource-node-id)
+                          renderable-meshes)]
+    {:node-id model-scene-resource-node-id
+     :transform transform
+     :aabb aabb
+     :children mesh-scenes}))
+
+(defn- make-scene [renderable-mesh-set model-scene-resource-node-id]
+  (let [{:keys [aabb renderable-models]} renderable-mesh-set
+        model-scenes (mapv #(make-model-scene % model-scene-resource-node-id)
+                           renderable-models)]
+    {:node-id model-scene-resource-node-id
+     :aabb aabb
+     :renderable {:render-fn render-outline
+                  :tags #{:model :outline}
+                  :batch-key nil ; Batching is disabled in the editor for simplicity.
+                  :select-batch-key model-scene-resource-node-id
+                  :passes [pass/outline]}
+     :children model-scenes}))
+
+(g/defnk produce-scene [_node-id renderable-mesh-set]
+  (make-scene renderable-mesh-set _node-id))
+
+(defn- augment-mesh-scene [mesh-scene old-node-id new-node-id new-node-outline-key material-name->material-scene-info]
+  (let [mesh-renderable (:renderable mesh-scene)
+        material-name (:material-name mesh-renderable)
+        material-scene-info (material-name->material-scene-info material-name)
+        claimed-scene (scene/claim-child-scene old-node-id new-node-id new-node-outline-key mesh-scene)]
```

Comment:
An issue with the old code was that the Model using a particular `model-scene` did not claim ownership of the scene-data produced by its `scene` output. This resulted in meshes that had different requirements in terms of materials and transforms sharing the same `request-id` for the `scene-cache`. There were also issues related to picking and selection, since the renderables could not be uniquely identified to belong to the referencing Model.